src/api-binder.coffee

```diff
@@ -0,0 +1,369 @@
+			.then =>
+				@config.get('initialConfigReported')
+				.then (reported) =>
+					if !checkTruthy(reported)
+						console.log('Reporting initial configuration')
+						@reportInitialConfig(bootstrapRetryDelay)
+			.then =>
+				console.log('Starting current state report')
+				@startCurrentStateReport()
+			.then =>
+				@readyForUpdates = true
+				console.log('Starting target state poll')
+				@startTargetStatePoll()
+			return
+
+	fetchDevice: (uuid, apiKey, timeout) =>
+		@resinApi.get
+			resource: 'device'
+			options:
+				filter:
+					uuid: uuid
+			customOptions:
+				apikey: apiKey
+		.get(0)
+		.catchReturn(null)
+		.timeout(timeout)
+
+	_exchangeKeyAndGetDevice: (opts) ->
+		Promise.try =>
+			if !opts?
+				@config.get('provisioningOptions')
+				.then (conf) ->
+					opts = conf
+		.then =>
+			# If we have an existing device key we first check if it's valid, because if it is we can just use that
+			if opts.deviceApiKey?
+				@fetchDevice(opts.uuid, opts.deviceApiKey, opts.apiTimeout)
+		.then (device) =>
+			if device?
+				return device
+			# If it's not valid/doesn't exist then we try to use the user/provisioning api key for the exchange
+			@fetchDevice(opts.uuid, opts.provisioningApiKey, opts.apiTimeout)
+			.then (device) ->
+				if not device?
+					throw new ExchangeKeyError("Couldn't fetch device with provisioning key")
+				# We found the device, we can try to register a working device key for it
+				request.postAsync("#{opts.apiEndpoint}/api-key/device/#{device.id}/device-key?apikey=#{opts.provisioningApiKey}", {
+					json: true
+					body:
+						apiKey: opts.deviceApiKey
+				})
+				.spread (res, body) ->
+					if res.statusCode != 200
+						throw new ExchangeKeyError("Couldn't register device key with provisioning key")
+				.timeout(opts.apiTimeout)
+				.return(device)
+
+	_exchangeKeyAndGetDeviceOrRegenerate: (opts) =>
+		@_exchangeKeyAndGetDevice(opts)
+		.tap ->
+			console.log('Key exchange succeeded, all good')
+		.tapCatch ExchangeKeyError, (err) =>
+			# If it fails we just have to reregister as a provisioning key doesn't have the ability to change existing devices
+			console.log('Exchanging key failed, having to reregister')
+			@config.regenerateRegistrationFields()
+
+	_provision: =>
+		@config.get('provisioningOptions')
+		.then (opts) =>
+			return if opts.registered_at? and opts.deviceId? and !opts.provisioningApiKey?
+			Promise.try ->
+				if opts.registered_at? and !opts.deviceId?
+					console.log('Device is registered but no device id available, attempting key exchange')
+					@_exchangeKeyAndGetDeviceOrRegenerate(opts)
+				else if !opts.registered_at?
+					console.log('New device detected. Provisioning...')
+					deviceRegister.register(opts)
+					.timeout(opts.apiTimeout)
+					.catch DuplicateUuidError, =>
+						console.log('UUID already registered, trying a key exchange')
+						@_exchangeKeyAndGetDeviceOrRegenerate(opts)
+					.tap ->
+						opts.registered_at = Date.now()
```

Comment:
what if the device has not NTP yet?

Reply:
We don't really care as this is actually a legacy value and no one really reads the actual datetime - it might as well just be a boolean saying whether the device is registered or not. But since previously registered devices from older versions will have this key, we maintain compatibility. If anyone wants to use this (which they shouldn't since we never documented its meaning), they should at least know that this is device time.